r/src/array_from_vector.cpp

```diff
@@ -970,6 +985,26 @@ bool CheckCompatibleFactor(SEXP obj, const std::shared_ptr<arrow::DataType>& typ
   return dict_type->value_type() == utf8();
 }
 
+arrow::Status CheckCompatibleList(SEXP obj,
+                                  const std::shared_ptr<arrow::DataType>& type) {
+  if (TYPEOF(obj) != VECSXP) {
+    return Status::RError("Conversion to array requires list input");
+  }
+
+  // All list elements must be the same type
+  std::shared_ptr<arrow::DataType> element_type = type -> value_type();
```

Comment:
Have attempted to move the validation logic to here, it's not working as expected yet 😅